nicer test feedback

sknorr opened the pull request:
@tomschr, I tried to improve the test runner a bit. Main new features:

* If there is a validation issue in good/, the error message will be written directly to the command line
* All validation errors (both for good/ and bad/) are written to a single file ("last-test-run-errors") which should make it easier to look at them
* All test cases are checked for well-formedness before being run against the RNG

The one main ugly tradeoff I made is that the test runner relies on xmllint/jing command line output now instead of on return codes. I don't think that is a practical issue though.

I didn't do anything about the NovDoc test runner. I suspect it would be easiest to symlink that to the GeekoDoc test runner.

(The commits are not super-clean topic-wise, sorry.)

tomschr commented:
@sknorr Thanks for your time and efforts! Much appreciated! 👍

I'm mostly fine with your changes, but I have one or two questions regarding the following items:

> The one main ugly tradeoff I made is that the test runner relies on xmllint/jing command line output now instead of on return codes. I don't think that is a practical issue though.

Why do you think this was necessary? Wouldn't be return codes better and are independent of any output? It worked before so I'm not sure why this was needed. 

> I didn't do anything about the NovDoc test runner. I suspect it would be easiest to symlink that to the GeekoDoc test runner.

No problem, I think this is not soo important. I would even don't spend any thoughts on testing Novdoc... 👻

tomschr commented on geekodoc/tests/run-tests.sh:
```diff
@@ -17,55 +17,52 @@ PROGDIR=${0%/*}
-
-function validator {
+function validator() {
     case "$VALIDATOR" in
```
Yes, that's probably better to have only one function for validation. The only thing that I would like to change is returning the return code (with `return $?`) as I've mentioned in my previous comment.

sknorr replied:
Yeah, I don't think you can get both error message as a variable and the return code too, unfortunately.

tomschr commented on geekodoc/tests/run-tests.sh:
```diff
@@ -139,17 +136,29 @@ function test_files() {
-       count_errors $GOOD_OR_BAD $result
-       loginfo "Validating '$xmlfile'... $resultstr"
+       loginfo -n "Validating '$xmlfile'... "
+       wellformedness=$(xmllint --noout --noent $xmlfile 2>&1)
```
A wellformedness check is really cool! I haven't thought about that! 👍

Wouldn't it be "better" to check for the return code and *not* the output? Something like:

```bash
xmllint --noout --noent $xmlfile 2>&1
if [ $? -ne 0 ]; then
# ....
```

sknorr replied:
Again, the way I am doing it, means I write the validation result into a variable, so I can't use the return code.

tomschr commented on geekodoc/tests/run-tests.sh:
```diff
@@ -190,20 +199,29 @@ while true ; do
                 ;;
             esac
             ;;
+        -n|--noremake)
```
You should add some help text also to the `print_help()` function. 😉

tomschr replied:
For some reason, I get this:

```
$ LANG=C ./run-tests.sh -n
run-tests.sh: invalid option -- 'n'
...
```

sknorr commented:
I have fixed the --help and cleaned up commits a bit. Should be mergeable now.

If you have an idea for how to continue using return codes while also not writing everything to file primarily, I am open to changes.

tomschr commented:
I think, we can merge it. Should we?
